llama-index-integrations/indices/llama-index-indices-managed-postgresml/llama_index/indices/managed/postgresml/base.py

```diff
@@ -0,0 +1,198 @@
+                    },
+                }
+            }
+        self.pipeline = Pipeline(pipeline_name, pipeline_schema)
+
+        # We must wrap self.collection.add_pipeline() with this async function
+        # This is a limitation of the pyo3 async implementation
+        async def add_pipeline():
+            await self.collection.add_pipeline(self.pipeline)
+
+        run_async_tasks([add_pipeline()])
+
+        if nodes:
+            self._insert(nodes)
+
+    def _insert(
+        self,
+        nodes: Sequence[BaseNode],
+        **insert_kwargs: Any,
+    ) -> None:
+        """Insert a set of documents (each a node)."""
+        documents = [
+            {
+                "id": node.node_id,
+                "content": node.get_content(),
+                "metadata": node.metadata,
+            }
+            for node in nodes
+        ]
+
+        args = {"parallel_batches": self.upsert_parallel_batches, **insert_kwargs}
+
+        # We must wrap self.collection.upsert_documents() with this async function
+        # This is a limitation of the pyo3 async implementation
+        async def upsert_documents():
+            await self.collection.upsert_documents(documents, args)
+
+        run_async_tasks([upsert_documents()])
+
+    def add_documents(
+        self,
+        docs: Sequence[Document],
+        **insert_kwargs: Any,
+    ) -> None:
+        nodes = [
+            TextNode(id_=doc.id_, text=doc.get_content(), metadata=doc.metadata)
+            for doc in docs
+        ]
+        self._insert(nodes, **insert_kwargs)
+
+    def delete_ref_doc(self, ref_doc_id: str) -> None:
+        # We must wrap self.collection.delete_documents() with this async function
+        # This is a limitation of the pyo3 async implementation
+        async def delete_documents():
+            await self.collection.delete_documents({"id": {"$eq": ref_doc_id}})
+
+        run_async_tasks([delete_documents()])
+
+    def update_ref_doc(self, document: Document) -> None:
+        node = TextNode(
+            id_=document.id_, text=document.get_content(), metadata=document.metadata
+        )
+        self._insert([node], merge=True)
+
+    def as_retriever(self, **kwargs: Any) -> BaseRetriever:
+        """Return a Retriever for this managed index."""
+        from llama_index.indices.managed.postgresml.retriever import (
+            PostgresMLRetriever,
+        )
+
+        return PostgresMLRetriever(self, **kwargs)
+
+    def as_query_engine(self, **kwargs: Any) -> BaseQueryEngine:
+        from llama_index.indices.managed.postgresml.retriever import (
+            PostgresMLRetriever,
+        )
+        from llama_index.indices.managed.postgresml.query import (
+            PostgresMLQueryEngine,
+        )
+
+        return PostgresMLQueryEngine(PostgresMLRetriever(self, **kwargs), **kwargs)
+
+    @classmethod
+    def from_documents(
+        cls: Type[IndexType],
+        collection_name: str,
```

Comment:
Mostly for consistency, but I was a little surprised to see `documents` not being the first arg. This will be a bit of a break from other interfaces (similar comment to the constructor and `nodes` not being the first arg)

Thoughts a making collection_name a kwarg so that we can modify the order? (And raise an error if collection_name isn't specified?)

No worries if not! Just a UX thing that came to mind here